### ISSUE TITLE
Get rid of `HaskellPrebuiltPackageInfo`

### ISSUE THREAD

**Issue by regnat** [tweag/rules_haskell#440]
I've been thinking a bit lately around #75, #367 and other related issues, and
my conclusion is that we should merge the `HaskellPrebuiltPackageInfo` and
`HaskellLibraryInfo` providers, but I'm not sure this is feasible in practice
so I want to gather a bit of feedback before trying to implement this.
Here's the interface I imagine:

# Low-level interface

We extend the `haskell_import` workspace rule to take as arguments
(more or less) all the arguments required by `HaskellLibraryInfo`, symlinks all
the required files into the workspace and returns a `HaskellLibraryInfo` for
the given package.

For example, instead of having say `haskell_import(name = "bytestring")`, we would
have something like

```python
haskell_import(
  name = "bytestring",
  package_id = "bytestring-0.10.8.2",
  version = "0.10.8.2",
  import_dirs = "/nix/store/a0dr5x063v0zm8z0z5fwh8dldg10yl07-ghc-8.4.3/lib/ghc-8.4.3/bytestring-0.10.8.2",
  header_files = ["fpstring.h"],
  boot_files = [],
  source_files = [],
  extra_source_files = [],
  ghc_args = [],
  exposed_modules = [
    "Data.ByteString Data.ByteString.Builder",
    "Data.ByteString.Builder.Extra Data.ByteString.Builder.Internal",
    "Data.ByteString.Builder.Prim Data.ByteString.Builder.Prim.Internal",
    "Data.ByteString.Char8 Data.ByteString.Internal Data.ByteString.Lazy",
    "Data.ByteString.Lazy.Builder Data.ByteString.Lazy.Builder.ASCII",
    "Data.ByteString.Lazy.Builder.Extras Data.ByteString.Lazy.Char8",
    "Data.ByteString.Lazy.Internal Data.ByteString.Short",
    "Data.ByteString.Short.Internal Data.ByteString.Unsafe",
  ],
)
```

which would Generate a `bytestring` repository containing:
- A symlink `import_dir` to `/nix/store/a0dr5x063v0zm8z0z5fwh8dldg10yl07-ghc-8.4.3/lib/ghc-8.4.3/bytestring-0.10.8.2`
- A file `exposed_modules_file` containing the list of `exposed_modules`
- A `BUILD` file, with a `package` rule which returns a `HaskellLibraryInfo`
  with the same arguments as the rule, except `import_dirs` which points
  to `@bytestring//:import_dir` and `exposed_modules_file` which is defined as
  `@bytestring//:exposed_modules_file`

# Higher-level interface

Obviously, these `haskell_import` rules wouldn't be written by hand, but rather be generated by some external tool or a workspace rule

- In the case of `hazel` I assume it's easy to extend the framework to this API.

- For the `Nix` case, we could either generate them from the `nixpkgs`
definitions, by having a `haskell_nix_import` rule with the same arguments as
`nix_package` which would generate such an `haskell_import` definition.

- Alternatively, we could also query `ghc-pkg` to generate such a rule for all
the available packages (which is what I've manually done for the example above).

# Advantages over the actual solution

This approach (if possible) has several advantages over the current seup:

- The inner code could (hopefully) be simplified since we wouldn't have to
  handle both cases of `HaskellLibraryInfo` and `HaskellPrebuiltPackageInfo`,
- This would bring a natural solution to bugs such as #367,
- In the Nix case, we wouldn't have to use `ghcWithPackages` anymore, which
  means that we wouldn't have to rebuild the whole repo each time a new
  dependency is added.

/cc @mboes @mrkkrp @guibou @judah , does this seem feasible/reasonable to you?

**Comment by lunaris** [comment]
I think this is a good call. I think the `haskell_nix_import` rule would complement some of the work I've been toying with in #387 and also perhaps be a nicer way of addressing several of the concerns in https://github.com/FormationAI/hazel/issues/41.

**Comment by Profpatsch** [comment]
https://github.com/tweag/rules_haskell/commit/1b3cabcb7a672b012868d57a967898136db34696 is in master, does that mean we can close this issue?

**Comment by mboes** [comment]
No, as pointed out in the PR associated with the commit you point to, #442 did not actually remove `HaskellPrebuiltPackageInfo`.

**Comment by mboes** [comment]
I have a branch (to be submitted as a PR shortly) that does the last mile, since #442 only got us part of the way there. However, it repurposes `haskell_import` entirely, meaning it's a breaking change. Specifically:

* `haskell_import` as we know it today is renamed to `haskell_toolchain_library`. The original name was a misnomer, since `cc_import` and `java_import` expect the actual build artifacts, rather than searching them somewhere in the toolchain using a global database.
* The new `haskell_import` is as in the description of this ticket. Actual static, dynamic libraries have to be provided, along with headers files and include paths if they exist. `haskell_import` returns all the same providers that a `haskell_library` does, including `CcInfo` wrapping the static and dynamic libraries. This resolves #838.

`haskell_import` is pretty pervasive in some projects, so this BC would be quite egregious. OTOH marking it as deprecated will generate a lot of deprecation spam, and adapting to the rename should be as simple as `sed -i s/haskell_import/haskell_toolchain_library/`. Thoughts?
